storage/s3.go

```diff
@@ -382,6 +386,175 @@ func (s *S3) Get(
 	})
 }
 
+type SelectRecord struct {
+	Record json.RawMessage
+	Err    error
+}
+
+func (s *S3) selectStream(resp *s3.SelectObjectContentOutput, resultch chan<- *SelectRecord) {
+	defer resp.EventStream.Close()
+
+	reader, writer := io.Pipe()
+	defer reader.Close()
+	go func() {
+		defer writer.Close()
+		// TODO: wire ctx up to eventStream.StreamCloser.Close
+		for event := range resp.EventStream.Reader.Events() {
+			switch e := event.(type) {
+			case *s3.RecordsEvent:
+				// TODO: how should we handle errors here?!
+				writer.Write(e.Payload)
+			// How the heck are these different?
+			case *s3.StatsEvent:
+			case *s3.ProgressEvent:
+			}
+		}
+	}()
+
+	decoder := json.NewDecoder(reader)
+	// decoder.DontCopyRawMessage()
+	for {
+		var record json.RawMessage
+		err := decoder.Decode(&record)
+		if err != nil {
+			if err != io.EOF {
+				resultch <- &SelectRecord{
+					Err: err,
+				}
+			}
+			return
+		}
+		resultch <- &SelectRecord{
+			Record: record,
+		}
+	}
+	err := resp.EventStream.Err()
+	if err != nil {
+		resultch <- &SelectRecord{
+			Err: err,
+		}
+	}
+}
+
+func (s *S3) selectWorker(ctx context.Context, input s3.SelectObjectContentInput, objch <-chan *Object, resultch chan<- *SelectRecord) {
+	for object := range objch {
+		// TODO: reintroduce
+		if object.Type.IsDir() { // || errorpkg.IsCancelation(object.Err) {
+			continue
+		}
+
+		if err := object.Err; err != nil {
+			resultch <- &SelectRecord{
+				Err: object.Err,
+			}
+			continue
+		}
+
+		if object.StorageClass.IsGlacier() {
+			// TODO: this isn't fatal - how to signal?
+			resultch <- &SelectRecord{
+				Err: fmt.Errorf("object '%v' is on Glacier storage", object),
+			}
+			continue
+		}
+
+		url := object.URL
+		input.Bucket = aws.String(url.Bucket)
+		input.Key = aws.String(url.Path)
+
+		out, err := s.api.SelectObjectContentWithContext(ctx, &input)
+		if err != nil {
+			resultch <- &SelectRecord{
+				Err: err,
+			}
+		} else {
+			s.selectStream(out, resultch)
+		}
+	}
+}
+
+type Query struct {
+	ExpressionType  string
+	Expression      string
+	CompressionType string
+}
+
+func writev(to io.Writer, vector ...[]byte) (err error) {
```

Comment:
A little misleading, seeing as it doesn't actually use the `writev` syscall.

Reply:
Will be unused after https://github.com/peak/s5cmd/pull/300#discussion_r654325903